### ISSUE TITLE
feature flags import from pygame.future feature?

### ISSUE THREAD

**Issue by illume** [pygame/pygame#3443]
There was a discussion on this started here: https://github.com/pygame/pygame/pull/3396#discussion_r950738414

**Comment by illume** [comment]
It's a good point that zoldalma999 made: people just using the function may not notice it's experimental, because they don't read the docs... they just use the function.

For this release I'll try something fairly low effort, because we want to get the release out soonish.

- Mark all new APIs as experimental. (luckily, they are labeled in the PRs)
- For each new API, I'll reopen the original issue for feedback... or create an issue if there wasn't one, linking to the PR.
- Link to the issue in the docs, and in the release notes.

This way:
- people can leave feedback
- the people who contributed to the new API will be able to get that feedback

**Comment by ankith26** [comment]
> There was another good comment on this topic somewhere else... but I forgot to link it and now can't find it.

Are you referring to [this](https://github.com/pygame/pygame/pull/3100#issuecomment-1221273545)?

About the `pygame.experimental` namespace, it works well when we are adding new experimental modules and functions, but not when we are adding new methods to pre-existing classes.

Anyways, I agree with all the points you mentioned above, and ATM our main problem is getting feedback from a wider range of users. The best way to get real user feedback is to actually release new API in the wilds while making their experimental nature clear enough. My idea for doing this is slightly different from feature flags imports, it is using warnings. We would have a new `pygame.ExperimentalAPIWarning` (or something of that sort) which would be a subclass of python `Warning`.

When we mark an API as experimental, we would also have to raise a descriptive warning to let users know it. Users may still choose to silence it, but that needs more conscious effort by the user and this way we are sure that the user is almost guaranteed to see the message atleast once.

**Comment by Starbuck5** [comment]
I don't think we need to implement such solutions for this problem.

The best way to manage API mistakes, in my opinion, would be on a case by case heuristic basis, with frequent releases.

Here are a few examples

I added a function `mouse.set_system_cursor`, and then superseded that work within a small amount of time (2 months?) with a more robust `mouse.set_cursor` API. That new API was promptly released, and the `set_system_cursor` function was deprecated. The biggest obstacle to the success of this maneuver was that the documentation wasn't updated for about a year, so people were greeted with deprecation warnings with no alternate path in the documentation.

I believe the scrap API needs to be fundamentally redone, and good news, the module is marked as experimental! Does that mean the pygame contributors have free reign to re-engineer the whole module? I'd say it doesn't, because despite it's experimental label that API has been stable for more than a decade, and changing it would break code.

In my view, whether we've marked it as experimental is much less important than real world effects.

Gating new APIs behind feature flags or alternative namespaces would reduce forwards compatibility for apps that use them, and reduce the number of people actually using them and giving feedback.

**Comment by Starbuck5** [comment]
Thank you for articulating those goals @illume, it's interesting to get out on "paper."

> - To avoid where things are implemented partially, and then finish/polish work is abandoned. We have a number of messes like this, and some of them have taken years and years to clean up.
> - We want to allow hacky, half-finished things so people can try out ideas.

Historically there has been a high bar to get your new [WHATEVER] merged into pygame, and I'm in favor of maintaining this. If newbies are flagging with an interesting feature, we can encourage collaborative branches, not every PR needs just one author. But I'm in favor of making sure _most_ loose ends are settled before merging into the main branch.

> - To not get stuck with a very bad API. Without people trying out an API it's often bad. Real world use with humans gives good feedback on if an API is good for real humans.
> - Allowing new APIs to be done quickly should be our goal.
> - bad APIs stay forever... or for quite long.

The best way to handle this in my opinion would be to maintain our new rolling deprecation cycle strategy, with newer features (i.e. less used) requiring less time before removal, and to increase the cadence of releases. Higher release frequency gets features out quicker, allows for more granularity with deprecations and documentation edits.

> - To signal to users that this API is experimental, and to get feedback from users. Hopefully collected in one place, perhaps in a single issue?

In the past people have not been shy about perceived API shortcomings. I believe we've all explained to someone that pygame can't completely reverse the way it does [X] because of backwards compatibility. Places like discord and reddit are good for getting feedback (more casually than issues), and we could actively invite API feedback on those platforms when we do releases. And of course people can open issues.

**Comment by illume** [comment]
> Anyways, I agree with all the points you mentioned above, and ATM our main problem is getting feedback from a wider range of users. The best way to get real user feedback is to actually release new API in the wilds while making their experimental nature clear enough. My idea for doing this is slightly different from feature flags imports, it is using warnings. We would have a new pygame.ExperimentalAPIWarning (or something of that sort) which would be a subclass of python Warning.
> 
> When we mark an API as experimental, we would also have to raise a descriptive warning to let users know it. Users may still choose to silence it, but that needs more conscious effort by the user and this way we are sure that the user is almost guaranteed to see the message atleast once.

Yes, I think warnings are the way to go. It's the best chance the user will get to see the API-experimental status. 

I'd add to that a link to an issue open for discussion (as mentioned before... an existing issue if there is one so previous participants are looped in). Because I think that's the best chance we'd get to gather feedback.

The issue link would appear in:
- the release notes
- the warning
- the API docs

**Comment by Starbuck5** [comment]
> I am very much in agreement @illume by letting new contributors hack on the game in a way that wouldn't pollute the namespace without explicit imports while making it obvious it may or may not be fully finished or there may be bugs, we are increasing the potential to get some amazing contributions from other people.

What I don't want is a bureaucratic process where every new feature doubles its time to market because it has to go through a mandatory experimental namespace phase.

And pygame already has plenty of unfinished and abandoned code throughout the codebase. Cython sprite, the entire _sdl2 system, for example. It can be done case by case, as is logical for the contribution.

Plus, even rejected PRs teach the authors something. Code not being upstreamed doesn't mean it was a waste. I wrote 3000 lines of C to port the entire _sdl2.video module from cython to C, and no one had the courage/wherewithal to review it, so I closed it myself after like 6 months. But I learned a ton.

**Comment by PurityLake** [comment]
> > I am very much in agreement @illume by letting new contributors hack on the game in a way that wouldn't pollute the namespace without explicit imports while making it obvious it may or may not be fully finished or there may be bugs, we are increasing the potential to get some amazing contributions from other people.
> 
> 
> 
> What I don't want is a bureaucratic process where every new feature doubles its time to market because it has to go through a mandatory experimental namespace phase.
> 
> 
> 
> And pygame already has plenty of unfinished and abandoned code throughout the codebase. Cython sprite, the entire _sdl2 system, for example. It can be done case by case, as is logical for the contribution.
> 
> 
> 
> Plus, even rejected PRs teach the authors something. Code not being upstreamed doesn't mean it was a waste. I wrote 3000 lines of C to port the entire _sdl2.video module from cython to C, and no one had the courage/wherewithal to review it, so I closed it myself after like 6 months. But I learned a ton.

You make some valid points, bureaucracy can be major turn off, usually means more busy work too. I do agree with illume, in my mind it makes sense to do so but maybe more so for larger new APIs
